src/framework/control.ml

```diff
@@ -20,10 +20,16 @@ let spec_module: (module Spec) Lazy.t = lazy (
   let open Batteries in
   (* apply functor F on module X if opt is true *)
   let lift opt (module F : S2S) (module X : Spec) = (module (val if opt then (module F (X)) else (module X) : Spec) : Spec) in
+  let module GasVal = struct
+    (* Chain lattice has elements [0,n-1], but we want [0,gas_value] *)
+    let n () = get_int "ana.context.gas_value" + 1
+  end
+  in
   let module S1 =
     (val
       (module MCP.MCP2 : Spec)
-      |> lift (get_int "ana.context.gas_value" >= 0) (module ContextGasLifter)
+      |> lift (get_int "ana.context.gas_value" >= 0 && get_string "ana.context.gas_scope" = "global") (module ContextGasLifter(GlobalGas(GasVal)))
+      |> lift (get_int "ana.context.gas_value" >= 0 && get_string "ana.context.gas_scope" = "function") (module ContextGasLifter(PerFunctionGas(GasVal)))
```

Comment:
One might be able to get rid of a bit of redundancy here by extracting a function that checks the `ana.context.gas_scope` configuration and returns the appropriate module.

Reply:
I managed to pull it out, but it was a bit of a fight, mainly because I wanted to still avoid constructing the module eagerly.
I'm not sure if this is the most elegant way, I'm not intricately familiar with first-class modules, feel free to suggest a more concise way to do this.

CC also @sim642: I think out of the Goblint team, you're probably the most familiar with first-class modules.